tests/basic/clocks.rs

```diff
@@ -0,0 +1,415 @@
+                    cond.notify_one();
+                });
+            }
+
+            let mut guard = lock.lock().unwrap();
+            while !*guard {
+                check_clock(|i, c| (c > 0) == (i == 0)); // spawned thread has not executed notify_one
+                guard = cond.wait(guard).unwrap();
+            }
+            check_clock(|i, c| (c > 0) == (i == 0 || i == 1));
+        },
+        None,
+    )
+}
+
+fn clock_condvar_notify_one(num_notifiers: usize, num_waiters: usize) {
+    let lock = Arc::new(Mutex::new(0usize));
+    let cond = Arc::new(Condvar::new());
+
+    for _ in 0..num_notifiers {
+        let lock = Arc::clone(&lock);
+        let cond = Arc::clone(&cond);
+        thread::spawn(move || {
+            assert!(me() <= num_notifiers);
+            *lock.lock().unwrap() = me();
+            cond.notify_one();
+        });
+    }
+
+    for _ in 0..num_waiters {
+        let lock = Arc::clone(&lock);
+        let cond = Arc::clone(&cond);
+        thread::spawn(move || {
+            let mut guard = lock.lock().unwrap();
+            while *guard == 0 {
+                check_clock(|i, c| !(i >= 1 && i <= num_notifiers) || (c == 0)); // no notifier has gone yet
+                guard = cond.wait(guard).unwrap();
+            }
+            // Note that since all the threads touch the lock, any of them may have preceded this thread.
+            // But we know for sure that the thread that unblocked us should causally precede us.
+            check_clock(|i, c| (i != *guard) || (c > 0));
+        });
+    }
+}
+
+#[test]
+fn clock_condvar_notify_one_dfs() {
+    check_dfs(|| clock_condvar_notify_one(1, 1), None);
+}
+
+#[test]
+fn clock_condvar_notify_one_pct() {
+    check_pct(|| clock_condvar_notify_one(10, 10), 10_000, 3);
+}
+
+fn clock_condvar_notify_all(num_waiters: usize) {
+    let lock = Arc::new(Mutex::new(0usize));
+    let cond = Arc::new(Condvar::new());
+
+    {
+        let lock = Arc::clone(&lock);
+        let cond = Arc::clone(&cond);
+        thread::spawn(move || {
+            assert_eq!(me(), 1);
+            *lock.lock().unwrap() = me();
+            cond.notify_all();
+        });
+    }
+
+    for _ in 0..num_waiters {
+        let lock = Arc::clone(&lock);
+        let cond = Arc::clone(&cond);
+        thread::spawn(move || {
+            let mut guard = lock.lock().unwrap();
+            while *guard == 0 {
+                check_clock(|i, c| (i != 1) || (c == 0)); // notifier hasn't been scheduled
+                guard = cond.wait(guard).unwrap();
+            }
+            // Note that since all the threads touch the lock, any of them may have preceded this thread.
+            // But we know for sure that the thread that unblocked us should causally precede us.
+            check_clock(|i, c| (i != *guard) || (c > 0));
+        });
+    }
+}
+
+#[test]
+fn clock_condvar_notify_all_dfs() {
+    check_dfs(|| clock_condvar_notify_all(2), None);
+}
```

Comment:
This test is really slow, is there any way to make it smaller?

Actually a bunch of the clock tests are really slow. Might be worth spending some time playing with the output of `cargo test -- -Zunstable-options --report-time` to pick off the slowest ones